common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,446 @@
+                .expect("Expected tokens")
+        );
+        rate_limiter
+            .acquire_tokens(1)
+            .expect_err("Expected time to wait");
+    }
+
+    // Check number of keys in the rate limiter
+    fn assert_num_keys(rate_limiters: &TokenBucketRateLimiter<&str>, num_keys: usize) {
+        assert_eq!(num_keys, rate_limiters.buckets.read().len())
+    }
+
+    macro_rules! NonZeroUsize {
+        ($n:expr) => {
+            NonZeroUsize::new($n).unwrap()
+        };
+    }
+
+    #[test]
+    fn test_rate_limiting() {
+        let bucket_size = 5;
+        let bucket_rate = 1;
+        let key = "Key";
+        let rate_limiter =
+            TokenBucketRateLimiter::new(NonZeroUsize!(bucket_size), NonZeroUsize!(bucket_rate));
+
+        let bucket_arc = rate_limiter.bucket(key);
+        let mut bucket = bucket_arc.lock();
+        assert_num_allowed(&mut bucket, bucket_size);
+
+        // After adding 2 token, only 2 should be allowed
+        bucket.add_tokens(2);
+        assert_num_allowed(&mut bucket, 2);
+
+        // Adding more tokens than the bucket size, should only give bucket size
+        bucket.add_tokens(bucket_size + 1);
+        assert_num_allowed(&mut bucket, bucket_size);
+    }
+
+    #[test]
+    fn test_message_rate_limiting() {
+        let bucket_size = 5;
+        let bucket_rate = 3;
+        let key = "Key";
+        let rate_limiter =
+            TokenBucketRateLimiter::new(NonZeroUsize!(bucket_size), NonZeroUsize!(bucket_rate));
+
+        let bucket_arc = rate_limiter.bucket(key);
+        let mut bucket = bucket_arc.lock();
+
+        // Larger than bucket, should never succeed
+        let result = bucket.acquire_all_tokens(bucket_size + 1);
+        assert!(result.expect_err("Should not give tokens").is_none());
+
+        // Normal case
+        let result = bucket.acquire_all_tokens(bucket_size);
+        result.expect("Should be successful");
+
+        // Test future wait
+        let result = bucket.acquire_all_tokens(bucket_size);
+        let wait_time = result
+            .expect_err("Should not succeed, but will in future")
+            .expect("Should have a time it succeeds");
+
+        sleep(wait_time.duration_since(Instant::now()));
+        let result = bucket.acquire_all_tokens(bucket_size);
+        result.expect("Should be successful");
+    }
+
+    #[test]
+    fn test_refill() {
+        let bucket_size = 5;
+        let bucket_rate = 1;
+        let key = "Key";
+        let rate_limiter =
+            TokenBucketRateLimiter::new(NonZeroUsize!(bucket_size), NonZeroUsize!(bucket_rate));
+
+        let bucket_arc = rate_limiter.bucket(key);
+        let mut bucket = bucket_arc.lock();
+        assert_num_allowed(&mut bucket, bucket_size);
+
+        // After 1 refill period, we should be at least 1 rate change if not more
+        // TODO: Put in a mock time service
+        sleep(bucket.time_of_next_refill().duration_since(Instant::now()));
```

Comment:
might be a good idea to adapt Philip's simulated time.

Reply:
Yeah, originally it hadn't been pushed through, but now it is so I can adapt it to it.